addons/web/controllers/json.py

```diff
@@ -244,6 +291,61 @@ def get_action_triples_():
         context.update(safe_eval(action.context, eval_context))
         return action, context, eval_context, record_id
 
+    def _web_json_2_rpc_readonly(self):
+        try:
+            model_name = request.httprequest.args['model']
+            method_name = request.httprequest.args['method']
+            Model = request.env[model_name]
+        except KeyError:
+            return False
+        for cls in Model.mro():
+            method = getattr(cls, method_name, None)
+            if method is not None and hasattr(method, '_readonly'):
+                return method._readonly
+        return False
+
+    @http.route(
+        '/json/2/<model>/<method>',
+        #methods=['POST'],  # must accept all methods, otherwise /json/<path:subpath> takes over
+        auth='bearer',
+        type='/json/2/rpc',
+        readonly=_web_json_2_rpc_readonly,
+        save_session=False
+    )
+    def web_json_2_rpc(
+        self,
+        model: str,
+        method: str,
+        ids: Sequence[int] = (),
+        kwargs: Mapping[str, Any] = frozendict(),
+        context: Mapping[str, Any] = frozendict(),
+    ):
+        if request.httprequest.method != 'POST':
+            raise MethodNotAllowed(response=request.make_json_response(
+                MethodNotAllowed.description, status=HTTPStatus.METHOD_NOT_ALLOWED))
+
+        try:
+            Model = request.env[model]
+        except KeyError as exc:
+            exc.error_response = request.make_json_response(
+                f"The model {model!r} does not exist", status=HTTPStatus.NOT_FOUND)
+            raise
+        # call_kw doesn't sanitize nor the ids, nor the context, so we don't
+        records = Model.with_context(context).browse(ids)
+        try:
+            func = get_public_method(records, method)
+        except AttributeError as exc:
+            exc.error_response = request.make_json_response(
+                exc.args[0], status=HTTPStatus.NOT_FOUND)
+            raise
+        signature = inspect.signature(func)
+        try:
+            signature.bind(records, **kwargs)
```

Comment:
In what case do you need to check this explicitly?

Reply:
I don't *need* it, it is only to provide a better error message to users.

I cannot do that:
```py
try:
    func(records, **kwargs)
except TypeError:
    ...  # user friendly message
```
because the TypeError would be too wide.

I can neither do:
```py
func(records, **kwargs)
```
because the error would bubble up, be intercepted by `except Exception` in `handle_error()` which hide the actual error.